internal/controllers/reference/indexer_test.go

```diff
@@ -0,0 +1,254 @@
+			require.NoError(t, err, "should not return error on setting reference")
+			_, exists, err := c.indexer.Get(&ObjectReference{Referrer: tc.checkReferrer, Referent: tc.checkReferent})
+			require.NoError(t, err)
+			require.Equal(t, tc.found, exists)
+		})
+	}
+}
+
+func TestObjectReferred(t *testing.T) {
+	testCases := []struct {
+		name          string
+		addReferrer   client.Object
+		addReferent   client.Object
+		checkReferent client.Object
+		referred      bool
+	}{
+		{
+			name:          "object_referred",
+			addReferrer:   testRefService1,
+			addReferent:   testRefSecret1,
+			checkReferent: testRefSecret1,
+			referred:      true,
+		},
+		{
+			name:          "object_not_referred",
+			addReferrer:   testRefService1,
+			addReferent:   testRefSecret1,
+			checkReferent: testRefSecret2,
+			referred:      false,
+		},
+	}
+
+	for _, tc := range testCases {
+		c := NewCacheIndexers()
+		err := c.SetObjectReference(tc.addReferrer, tc.addReferent)
+		require.NoError(t, err, "should not return error on setting reference")
+
+		referred, err := c.ObjectReferred(tc.checkReferent)
+		require.NoError(t, err)
+		require.Equal(t, tc.referred, referred)
+	}
+}
+
+func TestListReferredObjects(t *testing.T) {
+	testCases := []struct {
+		name          string
+		addReferrer   client.Object
+		addReferent   client.Object
+		checkReferrer client.Object
+		objectNum     int
+	}{
+		{
+			name:          "has_referred_objects",
+			addReferrer:   testRefService1,
+			addReferent:   testRefSecret1,
+			checkReferrer: testRefService1,
+			objectNum:     1,
+		},
+		{
+			name:          "has_no_referred_objects",
+			addReferrer:   testRefService1,
+			addReferent:   testRefSecret1,
+			checkReferrer: testRefService2,
+			objectNum:     0,
+		},
+	}
+
+	for _, tc := range testCases {
+		c := NewCacheIndexers()
+		err := c.SetObjectReference(tc.addReferrer, tc.addReferent)
+		require.NoError(t, err, "should not return error on setting reference")
+
+		referents, err := c.ListReferredObjects(tc.checkReferrer)
+		require.NoError(t, err)
+		require.Len(t, referents, tc.objectNum)
+	}
+}
+
+func TestDeleteReferencesByReferrer(t *testing.T) {
+	testCases := []struct {
+		name           string
+		deleteReferrer client.Object
+		checkReferrer  client.Object
+		checkReferent  client.Object
+		found          bool
+	}{
+		{
+			name:           "should_delete_references_with_referrer_correctly",
+			deleteReferrer: testRefService1,
+			checkReferrer:  testRefService1,
+			checkReferent:  testRefSecret1,
+			found:          false,
+		},
+		{
+			name:           "should_not_delete_unrelated_references",
+			deleteReferrer: testRefService1,
+			checkReferrer:  testRefService2,
+			checkReferent:  testRefSecret2,
+			found:          true,
+		},
+	}
+
+	for _, tc := range testCases {
+		c := NewCacheIndexers()
+		err := c.SetObjectReference(testRefService1, testRefSecret1)
+		require.NoError(t, err, "should not return error on setting reference")
```

Comment:
Just a note: `require` in here will break the whole test instead of only the test case. This might not be desirable and will break the "whole picture" of the test.

If I may, I'd suggest to use a subtest for this ( you already added the `name` variable which can be used to name the subtest ) + catching the test case variable `tc`. Similarly for other tests.